builtin/providers/azure/resource_azure_disk.go

```diff
@@ -0,0 +1,301 @@
+		}
+
+		// Wait until the data disk is attached
+		if err := mc.WaitForOperation(req, nil); err != nil {
+			return fmt.Errorf(
+				"Error waiting for data disk %s to be attached to instance %s: %s", d.Id(), vm, err)
+		}
+
+		// Make sure we return here since all possible changes are
+		// already updated if we reach this point
+		return nil
+	}
+
+	if d.HasChange("caching") || d.HasChange("lun") {
+		p := virtualmachinedisk.UpdateDataDiskParameters{
+			DiskName:    d.Id(),
+			Lun:         lun,
+			HostCaching: hostCaching(d),
+			MediaLink:   mediaLink(d),
+		}
+
+		log.Printf("[DEBUG] Updating data disk: %s", d.Id())
+		req, err := virtualmachinedisk.NewClient(mc).UpdateDataDisk(vm, vm, vm, lun, p)
+		if err != nil {
+			return fmt.Errorf("Error updating data disk %s: %s", d.Id(), err)
+		}
+
+		// Wait until the data disk is updated
+		if err := mc.WaitForOperation(req, nil); err != nil {
+			return fmt.Errorf(
+				"Error waiting for data disk %s to be updated: %s", d.Id(), err)
+		}
+	}
+
+	return resourceAzureDataDiskRead(d, meta)
+}
+
+func resourceAzureDataDiskDelete(d *schema.ResourceData, meta interface{}) error {
+	mc := meta.(management.Client)
+	lun := d.Get("lun").(int)
+	vm := d.Get("virtual_machine").(string)
+
+	// If a name was not supplied, it means we created a new emtpy disk and we now want to
+	// delete that disk again. Otherwise we only want to detach the disk and keep the blob.
+	_, removeBlob := d.GetOk("name")
+
+	log.Printf("[DEBUG] Detaching data disk %s with removeBlob = %t", d.Id(), removeBlob)
+	req, err := virtualmachinedisk.NewClient(mc).DeleteDataDisk(vm, vm, vm, lun, removeBlob)
+	if err != nil {
+		return fmt.Errorf(
+			"Error detaching data disk %s with removeBlob = %t: %s", d.Id(), removeBlob, err)
+	}
+
+	// Wait until the data disk is detached and optionally deleted
+	if err := mc.WaitForOperation(req, nil); err != nil {
+		return fmt.Errorf(
+			"Error waiting for data disk %s to be detached with removeBlob = %t: %s",
+			d.Id(), removeBlob, err)
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func hostCaching(d *schema.ResourceData) virtualmachinedisk.HostCachingType {
+	switch d.Get("caching").(string) {
+	case "ReadOnly":
+		return virtualmachinedisk.HostCachingTypeReadOnly
+	case "ReadWrite":
+		return virtualmachinedisk.HostCachingTypeReadWrite
+	default:
+		return virtualmachinedisk.HostCachingTypeNone
+	}
+}
+
+func mediaLink(d *schema.ResourceData) string {
+	mediaLink, ok := d.GetOk("media_link")
+	if ok {
+		return mediaLink.(string)
+	}
+
+	name, ok := d.GetOk("name")
+	if !ok {
+		name = fmt.Sprintf("%s-%d", d.Get("virtual_machine").(string), d.Get("lun").(int))
+	}
+
+	return fmt.Sprintf(dataDiskBlobStorageURL, d.Get("storage").(string), name.(string))
+}
+
+func verifyDataDiskParameters(d *schema.ResourceData) error {
+	caching := d.Get("caching").(string)
+	if caching != "None" && caching != "ReadOnly" && caching != "ReadWrite" {
```

Comment:
If you cast to ```HostCachingType``` you should be able to do a simple switch here:

switch caching{
case HostCachingTypeNone, HostCachingTypeReadOnly, HostCachingTypeReadWrite:
default:
    return fmt.Errorf()
}

I generally avoid using magic strings. Using constants and variables avoids the need to do a search and replace if the authors of the 3rd party module decide to change those strings for whatever reason.